### PR TITLE
Add compatibility with `league/oauth2-server:^9`

### PR DESCRIPTION
**Pull request by ajgarlag**

### CODE REVIEW

**Comment by wmouwen** [comment]
Would love to see this PR to be released when stable! :rocket:

**Comment by chalasr** [comment]
@ajgarlag would you appreciate some help finishing this? thanks for the PR!

**Comment by ajgarlag** [comment]
> @ajgarlag would you appreciate some help finishing this? thanks for the PR!

@chalasr Sure! Any help is welcome! 

I think that there is a supported use case that we should deprecate: calling the authorization controller without any authenticated user (in other PR). WDYT?

**Comment by chalasr** [comment]
> Sure! Any help is welcome!

Is there anything missing actually? I just see WIP in the title but this looks good

**Comment by ajgarlag** [comment]
> > Sure! Any help is welcome!
> 
> Is there anything missing actually? I just see WIP in the title but this looks good

My main concern is with the last commit, but I think this will affect only a minimal number of users.

**Comment by ajgarlag** [comment]
> > I think that there is a supported use case that we should deprecate: calling the authorization controller without any authenticated user (in other PR). WDYT?
> 
> I agree. We should also disallow passing `null` to `UserConverterInterface::toLeague()` isn't it?

I would merge this PR and disallow passing `null` to `UserConverterInterface::toLeague()` in a new PR before tagging a new version. WDYT?

Does Symfony allow an empty string as a user identifier?

**Comment by chalasr** [comment]
> I would merge this PR and disallow passing null to UserConverterInterface::toLeague() in a new PR before tagging a new version. WDYT?

Works for me 👍 

> Does Symfony allow an empty string as a user identifier?

No, built-in authenticators forbid it https://github.com/symfony/symfony/blob/dd9fa153d344593c7fe12eb761158558377d98a4/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L149

File: src/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -31,6 +32,11 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->defaultValue('ROLE_OAUTH2_')
                     ->cannotBeEmpty()
                 ->end()
+                ->scalarNode('anonymous_user_identifier')
```
Comment (chalasr) [inline src/DependencyInjection/Configuration.php:35]: Can you please remind me under which flow do we expect `anonymous` users? Wording-wise the `anonymous` concept has been removed in Symfony but I feel like this one is different and specific to this bundle, what's the case I'm missing?

Comment (ajgarlag) [reply src/DependencyInjection/Configuration.php:35]: The `AuthorizationController` does not require authentication (although it is highly recommended), so when the `AuthorizationRequestResolveEvent` is dispatched, there is a valid use case where the `AuthorizationRequestUserResolvingListener` cannot set the user instance in the event:
https://github.com/thephpleague/oauth2-server-bundle/blob/ec2c2d3a13f1f4b02de4c0472f0b5bc3bd1db33a/src/EventListener/AuthorizationRequestUserResolvingListenerTrait.php#L12-L18

So the `$event->getUser()` call can return null with a resolved authorization https://github.com/thephpleague/oauth2-server-bundle/blob/ec2c2d3a13f1f4b02de4c0472f0b5bc3bd1db33a/src/Controller/AuthorizationController.php#L104

This is the use case I think we should disallow.

**Comment by ajgarlag** [comment]
> > I would merge this PR and disallow passing null to UserConverterInterface::toLeague() in a new PR before tagging a new version. WDYT?
> 
> Works for me 👍
> 
> > Does Symfony allow an empty string as a user identifier?
> 
> No, built-in authenticators forbid it https://github.com/symfony/symfony/blob/dd9fa153d344593c7fe12eb761158558377d98a4/src/Symfony/Component/Security/Http/Authenticator/JsonLoginAuthenticator.php#L149

But the interface allows it. Maybe a `/** @return non-empty-string */` annotation should be added to the `UserInterface`. I'll submit a PR to deprecate this behavior in Symfony 7

**Comment by ajgarlag** [comment]
> But the interface allows it. Maybe a `/** @return non-empty-string */` annotation should be added to the `UserInterface`. I'll submit a PR to deprecate this behavior in Symfony 7

See https://github.com/symfony/symfony/pull/58007

File: src/Repository/NullAccessTokenRepository.php [hunk 2]
```diff
@@ -11,12 +11,13 @@
...
 {
-    public function getNewToken(ClientEntityInterface $clientEntity, array $scopes, $userIdentifier = null): AccessTokenEntityInterface
+    public function getNewToken(ClientEntityInterface $clientEntity, array $scopes, ?string $userIdentifier = null): AccessTokenEntityInterface
     {
-        /** @var int|string|null $userIdentifier */
         $accessToken = new AccessTokenEntity();
         $accessToken->setClient($clientEntity);
-        $accessToken->setUserIdentifier($userIdentifier);
+        if (null !== $userIdentifier && '' !== $userIdentifier) {
+            $accessToken->setUserIdentifier($userIdentifier);
```
Comment (chalasr) [inline src/Repository/NullAccessTokenRepository.php:19]: Given the low potential of breaking people's code as you said + the fact we're still in 0.x version (not for long I think!), I'm fine doing this change although it's a bit aggressive.

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Thank you @ajgarlag.

**Pull request merged**